sdks/python/apache_beam/metrics/cells.py

```diff
@@ -0,0 +1,305 @@
+
+  A MetricCell represents a specific metric in a single context and bundle.
+  All subclasses must be thread safe, as these are used in the pipeline runners,
+  and may be subject to parallel/concurrent updates. Cells should only be used
+  directly within a runner.
+  """
+  def __init__(self):
+    self.commit = CellCommitState()
+    self._lock = threading.Lock()
+
+  def get_cumulative(self):
+    raise NotImplementedError
+
+
+class CounterCell(Counter, MetricCell):
+  """Tracks the current value and delta of a counter metric.
+
+  Each cell tracks the state of a metric independently per context per bundle.
+  Therefore, each metric has a different cell in each bundle, cells are
+  aggregated by the runner.
+
+  This class is thread safe.
+  """
+  def __init__(self, *args):
+    super(CounterCell, self).__init__(*args)
+    self.value = 0
+
+  def combine(self, other):
+    result = CounterCell()
+    result.inc(self.value + other.value)
+    return result
+
+  def inc(self, n=1):
+    with self._lock:
+      self.value += n
+      self.commit.after_modification()
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.value
+
+
+class DistributionCell(Distribution, MetricCell):
+  """Tracks the current value and delta for a distribution metric.
+
+  Each cell tracks the state of a metric independently per context per bundle.
+  Therefore, each metric has a different cell in each bundle, that is later
+  aggregated.
+
+  This class is thread safe.
+  """
+  def __init__(self, *args):
+    super(DistributionCell, self).__init__(*args)
+    self.data = DistributionData(0, 0, None, None)
+
+  def combine(self, other):
+    result = DistributionCell()
+    result.data = self.data.combine(other.data)
+    return result
+
+  def update(self, value):
+    with self._lock:
+      self.commit.after_modification()
+      self._update(value)
+
+  def _update(self, value):
+    value = int(value)
+    self.data.count += 1
+    self.data.sum += value
+    self.data.min = (value
+                     if self.data.min is None or self.data.min > value
+                     else self.data.min)
+    self.data.max = (value
+                     if self.data.max is None or self.data.max < value
+                     else self.data.max)
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.data.get_cumulative()
+
+
+class DistributionResult(object):
+  """The result of a Distribution metric.
+  """
+  def __init__(self, data):
+    self.data = data
+
+  def __eq__(self, other):
+    return self.data == other.data
+
+  @property
+  def max(self):
+    return self.data.max
+
+  @property
+  def min(self):
+    return self.data.min
+
+  @property
+  def count(self):
+    return self.data.count
+
+  @property
+  def sum(self):
+    return self.data.sum
+
+  @property
+  def mean(self):
+    """Returns the float mean of the distribution."""
+    return float(self.data.sum)/self.data.count
```

Comment:
this will raise an error if no elements are in the distribution...not sure what the correct behavior should be though